send-transaction-service/src/send_transaction_service.rs

```diff
@@ -136,70 +137,88 @@ impl SendTransactionService {
 
         Builder::new()
             .name("send-tx-sv2".to_string())
-            .spawn(move || loop {
-                match receiver.recv_timeout(Duration::from_millis(1000.min(config.retry_rate_ms))) {
-                    Err(RecvTimeoutError::Disconnected) => break,
-                    Err(RecvTimeoutError::Timeout) => {}
-                    Ok(transaction_info) => {
-                        inc_new_counter_info!("send_transaction_service-recv-tx", 1);
-                        let addresses = leader_info.as_ref().map(|leader_info| {
-                            leader_info.get_leader_tpus(config.leader_forward_count)
-                        });
-                        let addresses = addresses
-                            .map(|address_list| {
-                                if address_list.is_empty() {
-                                    vec![&tpu_address]
-                                } else {
-                                    address_list
-                                }
-                            })
-                            .unwrap_or_else(|| vec![&tpu_address]);
-                        for address in addresses {
-                            Self::send_transaction(
-                                &send_socket,
-                                address,
-                                &transaction_info.wire_transaction,
+            .spawn(move || {
+                //if retry_rate_ms is high the expecation is that clients do their own retry logic and may spam the node
+                //faster then once very 400ms.
+                //if retry_rate_ms is low this stops retries from going faster then once every 400ms.
+                const MAX_DEDUPER_AGE: Duration = Duration::from_millis(400);
+                const MAX_DEDUPER_ITEMS: u32 = 1_000_000;
+                let mut deduper = Deduper::new(MAX_DEDUPER_ITEMS, MAX_DEDUPER_AGE);
+                loop {
+                    deduper.reset();
+                    match receiver
+                        .recv_timeout(Duration::from_millis(1000.min(config.retry_rate_ms)))
+                    {
+                        Err(RecvTimeoutError::Disconnected) => break,
+                        Err(RecvTimeoutError::Timeout) => {}
+                        Ok(transaction_info) => {
+                            inc_new_counter_info!("send_transaction_service-recv-tx", 1);
+                            let addresses = leader_info.as_ref().map(|leader_info| {
+                                leader_info.get_leader_tpus(config.leader_forward_count)
```

Comment:
I bet this is a cause of duplicates as well. The RPC node uses its local poh clock to determine who the upcoming leaders are, but if the RPC server falls behind the cluster, it's going to be sending copies of transactions to non-leader nodes which will all forward to the actual leader.

So I think we should change from local poh to use some heuristic utilizing the info from `MaxSlots` to get an accurate picture of the actual upcoming leaders even if the RPC node is behind.

It'd be nice to put a max on `leader_forward_count` as well